bottlerocket-agents/src/bin/eks-resource-agent/eks_provider.rs

```diff
@@ -644,53 +644,73 @@ async fn security_group(
     Ok(security_groups)
 }
 
-async fn instance_profile(
+async fn instance_profile_arn(
     iam_client: &aws_sdk_iam::Client,
-    cluster_name: &str,
-    node_instance_role: &str,
+    role_arn: &str,
 ) -> ProviderResult<String> {
-    let list_result = iam_client
-        .list_instance_profiles()
+    let role_name = role_name_from_arn(role_arn, Resources::Remaining)?;
+    let instance_profiles = iam_client
+        .list_instance_profiles_for_role()
+        .role_name(role_name)
         .send()
         .await
-        .context(Resources::Remaining, "Unable to list instance profiles")?;
-    let eksctl_prefix = format!("eksctl-{}", cluster_name);
-    list_result
-        .instance_profiles
-        .as_ref()
-        .context(Resources::Remaining, "No instance profiles found")?
-        .iter()
-        .filter(|x| {
-            x.instance_profile_name
-                .as_ref()
-                .unwrap_or(&"".to_string())
-                .contains("NodeInstanceProfile")
-        })
-        .filter(|x| {
-            x.instance_profile_name
-                .as_ref()
-                .unwrap_or(&"".to_string())
-                .contains(&eksctl_prefix)
-        })
-        .find(|instance_profile| {
-            instance_profile
-                .roles
-                .as_ref()
-                .map(|roles| {
-                    roles
-                        .iter()
-                        .any(|role| role.arn == Some(node_instance_role.to_string()))
-                })
-                .unwrap_or_default()
-        })
-        .context(Resources::Remaining, "Node instance profile not found")?
-        .arn
-        .as_ref()
         .context(
             Resources::Remaining,
-            "Node instance profile missing arn field",
+            format!(
+                "Unable to list instance profiles for role name '{}' \
+                which was derived from role arn '{}'",
+                role_name, role_arn
+            ),
+        )?
+        .instance_profiles
+        .ok_or_else(|| {
+            ProviderError::new_with_context(
+                Resources::Remaining,
+                "Instance profile list is missing from list_instance_profiles_for_role response",
+            )
+        })?;
+
+    if instance_profiles.len() > 1 {
+        return Err(ProviderError::new_with_context(
+            Resources::Remaining,
+            format!(
+                "More than one instance profile was found for role name \
+                '{}' which was derived from arn '{}'",
+                role_name, role_arn
+            ),
+        ));
+    }
+
+    let instance_profile = instance_profiles.into_iter().next().ok_or_else(|| {
+        ProviderError::new_with_context(
+            Resources::Remaining,
+            format!(
+                "More than one instance profile was found for role name \
+                '{}' which was derived from arn '{}'",
+                role_name, role_arn
+            ),
```

Comment:
This should mean there's no instance profile for the role at all.

```suggestion
    let instance_profile = instance_profiles.into_iter().next().ok_or_else(|| {
        ProviderError::new_with_context(
            Resources::Remaining,
            format!(
                "No instance profile was found for role '{}', role_arn
            ),
```

Reply:
Good catch.